src/test/java/com/aws/greengrass/clientdevices/auth/certificate/ClientCertificateStoreTest.java

```diff
@@ -0,0 +1,115 @@
+        String certPem = "pem";
+        String certId = "id";
+
+        assertThat(certificateStore.exists(certId), is(false));
+        certificateStore.storePem(certId, certPem);
+        assertThat(certificateStore.exists(certId), is(true));
+
+        Optional<String> returnedPem = certificateStore.getPem(certId);
+        assertThat(returnedPem.isPresent(), is(true));
+        assertThat(returnedPem.get(), is(certPem));
+
+        // Count files in work dir
+        try (Stream<Path> stream = Files.walk(workDir)) {
+            long count = stream.filter(Files::isRegularFile)
+                    .count();
+            assertThat(count, is(1L));
+        }
+    }
+
+    @Test
+    void GIVEN_multipleClientCertificates_WHEN_getPem_THEN_correctCertificateReturned() throws IOException {
+        Function<Integer, String> getId = (i) -> "id-" + i;
+        Function<Integer, String> getPem = (i) -> "pem-" + i;
+        long certCount = 10;
+
+        for (int i = 0; i < certCount; i++) {
+            assertThat(certificateStore.exists(getId.apply(i)), is(false));
+            certificateStore.storePem(getId.apply(i), getPem.apply(i));
+        }
+
+        for (int i = 0; i < certCount; i++) {
+            String id = getId.apply(i);
+
+            assertThat(certificateStore.exists(id), is(true));
+            Optional<String> returnedPem = certificateStore.getPem(id);
+            assertThat(returnedPem.isPresent(), is(true));
+            assertThat(returnedPem.get(), is(getPem.apply(i)));
+        }
+
+        // Count files in work dir
+        try (Stream<Path> stream = Files.walk(workDir)) {
+            long count = stream.filter(Files::isRegularFile).count();
+            assertThat(count, is(certCount));
+        }
+    }
+
+    @Test
+    void GIVEN_workDirectoryWithClientCertificates_WHEN_getPem_THEN_correctCertificateReturned() throws IOException {
+        Function<Integer, String> getId = (i) -> "id-" + i;
+        Function<Integer, String> getPem = (i) -> "pem-" + i;
+        long certCount = 10;
+
+        // GIVEN certificates stored in work directory
+        for (int i = 0; i < certCount; i++) {
+            assertThat(certificateStore.exists(getId.apply(i)), is(false));
+            certificateStore.storePem(getId.apply(i), getPem.apply(i));
+        }
+
+        // Simulate restart by re-creating certificate store from work dir
+        certificateStore = new ClientCertificateStore(workDir);
+
+        for (int i = 0; i < certCount; i++) {
+            String id = getId.apply(i);
+
+            assertThat(certificateStore.exists(id), is(true));
+            Optional<String> returnedPem = certificateStore.getPem(id);
+            assertThat(returnedPem.isPresent(), is(true));
+            assertThat(returnedPem.get(), is(getPem.apply(i)));
+        }
+
+        // Count files in work dir
+        try (Stream<Path> stream = Files.walk(workDir)) {
```

Comment:
Not a bad test, but this is an assertion that is easily bypassed if you just write the same number of random files with random content into the test folder.  A better assertion is to get the file by the certificate id and check the content matches. Or if we are exposing too many details about the implementation to the test we could completely remove this

Reply:
This is essentially just the above test, but after a reload. The loop above does check retrieved certs against the expected value.

It may actually make sense for this test to be coupled with the implementation (e.g. know details of how certs are stored, and assert that they are written where we expect them to be written). The reason is because we could completely change the file format on disk and these tests would still pass. Whereas that would "soft break" a customer who upgraded.

Let's revisit in a followup